File: golang/lc-238.go
package main

func productExceptSelf(nums []int) []int {
	leftProducts, rightProducts := make([]int, len(nums)), make([]int, len(nums))
	val := 1
	for i := 0; i < len(nums); i++ {
		val *= nums[i]
		leftProducts[i] = val
	}
	val = 1
	for i := len(nums) - 1; i >= 0; i-- {
		val *= nums[i]
		rightProducts[i] = val
	}

	res := make([]int, len(nums))
	res[0] = rightProducts[1]
	res[len(res)-1] = leftProducts[len(res)-2]
	for i := 1; i < len(nums)-1; i++ {
		res[i] = rightProducts[i+1] * leftProducts[i-1]
	}
	return res
}

File: js&ts/LeetCode75/lc-238.js
/**
 * @param {number[]} nums
 * @return {number[]}
 */
var productExceptSelf = function (nums) {
  let left = new Array(nums.length);
  let right = new Array(nums.length);
  let val = 1;
  for (let i = 0; i < nums.length; ++i) {
    val *= nums[i];
    left[i] = val;
  }
  val = 1;
  for (let i = nums.length - 1; i >= 0; --i) {
    val *= nums[i];
    right[i] = val;
  }
  let res = Array(nums.length);
  res[0] = right[1];
  res[nums.length - 1] = left[nums.length - 2];
  for (let i = 1; i < nums.length - 1; ++i) {
    res[i] = left[i - 1] * right[i + 1];
  }
  return res;
};
